go-controller/pkg/clustermanager/routeadvertisements/controller.go

```diff
@@ -0,0 +1,1145 @@
+		// Prepare each scenario with a switch statement and check after that
+		switch {
+		case targetVRF == "auto" && router.VRF == "":
+			// match on default network/VRF, advertise node prefixes and receive
+			// any prefix of default network.
+			matchedVRF = ""
+			matchedNetwork = types.DefaultNetworkName
+			advertisePrefixes = selectedNetworks.hostNetworkSubnets[matchedNetwork]
+			receivePrefixes = selectedNetworks.networkSubnets[matchedNetwork]
+		case targetVRF == "auto":
+			// match router.VRF to network.VRF, advertise node prefixes and
+			// receive any prefix of the matched network
+			matchedVRF = router.VRF
+			matchedNetwork = selectedNetworks.networkVRFs[matchedVRF]
+			advertisePrefixes = selectedNetworks.hostNetworkSubnets[matchedNetwork]
+			receivePrefixes = selectedNetworks.networkSubnets[matchedNetwork]
+		case targetVRF == "":
+			// match on default network/VRF, advertise node prefixes and
+			// receive any prefix of selected networks
+			matchedVRF = ""
+			matchedNetwork = types.DefaultNetworkName
+			advertisePrefixes = selectedNetworks.hostSubnets
+			receivePrefixes = selectedNetworks.subnets
+		default:
+			// match router.VRF to network.VRF, advertise node prefixes and
+			// receive any prefix of selected networks
+			matchedVRF = targetVRF
+			matchedNetwork = selectedNetworks.networkVRFs[matchedVRF]
+			advertisePrefixes = selectedNetworks.hostSubnets
+			receivePrefixes = selectedNetworks.subnets
+		}
+		if matchedVRF != router.VRF || len(advertisePrefixes) == 0 {
+			// either this router VRF does not match the target VRF or we don't
+			// have prefixes for it (which might be due to this RA not selecting
+			// this network, but not just)
+			continue
+		}
+		matchedNetworks.Insert(matchedNetwork)
+
+		// if this router's VRF matches the target VRF, copy it and set the
+		// prefixes as appropriate
+		targetRouter := router
+		targetRouter.Prefixes = advertisePrefixes
+		targetRouter.Neighbors = make([]frrtypes.Neighbor, 0, len(source.Spec.BGP.Routers[i].Neighbors))
+		for _, neighbor := range source.Spec.BGP.Routers[i].Neighbors {
+			advertisePrefixes := advertisePrefixes
+			receivePrefixes := receivePrefixes
+			if neighbor.DisableMP {
+				isIPV6 := utilnet.IsIPv6String(neighbor.Address)
+				advertisePrefixes = util.MatchAllIPNetsStringFamily(isIPV6, advertisePrefixes)
+				receivePrefixes = util.MatchAllIPNetsStringFamily(isIPV6, receivePrefixes)
+			}
+			if len(advertisePrefixes) == 0 {
+				continue
+			}
+			neighbor.ToAdvertise = frrtypes.Advertise{
+				Allowed: frrtypes.AllowedOutPrefixes{
+					Mode:     frrtypes.AllowRestricted,
+					Prefixes: advertisePrefixes,
+				},
+			}
+			neighbor.ToReceive = frrtypes.Receive{
+				Allowed: frrtypes.AllowedInPrefixes{
+					Mode: frrtypes.AllowRestricted,
+				},
+			}
+			if advertisements.Has(ratypes.PodNetwork) {
+				for _, prefix := range receivePrefixes {
+					neighbor.ToReceive.Allowed.Prefixes = append(neighbor.ToReceive.Allowed.Prefixes,
```

Comment:
Why we accept routes to node subnets? Do we want to steer the host-to-pod traffic via br-ex instead of mp0?

Reply:
If pod-to-host goes through br-ex, then we need the replies to follow a symmetric path.
There's work ongoing to change pod-to-host to go through the overlay but has not landed yet.